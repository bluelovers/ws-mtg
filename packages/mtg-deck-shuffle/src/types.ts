import { Random } from 'random-extra/src';
import { ICardOfLibrary } from 'mtg-decklist-to-library';
import { IOptions as IOptionsDecklistToLibrary } from 'mtg-decklist-to-library';

export interface IOptionsDeckLibraryWithShuffle<T = ICardOfLibrary> extends IOptionsDecklistToLibrary<T>
{
	maxChunkLength?: number,
	random?: Random
}
