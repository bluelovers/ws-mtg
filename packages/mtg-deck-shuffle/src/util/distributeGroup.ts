import { ICardOfLibraryBase } from 'mtg-decklist-to-library';
import { createGroupArray } from './createGroupArray';

export function distributeGroup<T extends ICardOfLibraryBase = ICardOfLibraryBase>(group: Record<string, T[]>)
{
	let arr = createGroupArray<T>(4);

	const names = Object.keys(group);

	let card: T;
	let i = arr.length - 1;

	do
	{

		for (let name of names)
		{
			card = group[name]?.pop();

			if (card)
			{
				arr[i % arr.length].push(card);
			}
			else
			{
				delete group[name]
			}
		}

		if (card)
		{
			i++;
		}

	}
	while (Object.keys(group).length)

	return arr.flat()
}
