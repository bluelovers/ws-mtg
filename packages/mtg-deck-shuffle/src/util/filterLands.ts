import { ICardWithoutAmount } from 'mtg-decklist-parser2';
import { parseSnowCoveredOrBaseLand } from 'mtg-base-land';

export function filterLands<T extends ICardWithoutAmount = ICardWithoutAmount>(cards: T[])
{
	return cards.reduce((data, card) =>
	{

		let info = parseSnowCoveredOrBaseLand(card.name);

		if (!info)
		{
			data.others.push(card);
		}
		else if (info?.snow)
		{
			data.snowLands.push(card);
		}
		else
		{
			data.baseLands.push(card);
		}

		return data
	}, {
		baseLands: [] as T[],
		snowLands: [] as T[],
		others: [] as T[],
	})
}
