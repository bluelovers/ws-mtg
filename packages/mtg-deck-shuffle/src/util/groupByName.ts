import { ICardOfLibraryBase } from 'mtg-decklist-to-library';

export function groupByName<T extends ICardOfLibraryBase = ICardOfLibraryBase>(cards: T[])
{
	return cards.reduce((map, card) =>
	{

		map[card.name] ??= [];
		map[card.name].push(card);

		return map
	}, {} as Record<string, T[]>);
}
