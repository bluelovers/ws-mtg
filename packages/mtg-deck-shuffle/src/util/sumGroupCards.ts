import { ICardOfLibraryBase } from 'mtg-decklist-to-library';

export function sumGroupCards<T = ICardOfLibraryBase>(group: Record<string, T[]>)
{
	return Object.values(group).reduce((a, b) => a + b.length, 0)
}
