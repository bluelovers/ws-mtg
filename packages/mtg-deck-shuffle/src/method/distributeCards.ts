import { ICardOfLibraryBase } from 'mtg-decklist-to-library';
import { IOptionsDeckLibraryWithShuffle } from '../types';
import { filterLands } from '../util/filterLands';
import { groupByName } from '../util/groupByName';
import { distributeGroup } from '../util/distributeGroup';
import { createGroupArray } from '../util/createGroupArray';

export function distributeCards<T extends ICardOfLibraryBase = ICardOfLibraryBase>(cards: T[],
	// @ts-ignore
	options?: IOptionsDeckLibraryWithShuffle
): T[]
{
	let data = filterLands(cards);

	let arr = createGroupArray<T>(4);

	let idx = 0;

	data.baseLands = distributeGroup(groupByName(data.baseLands));
	data.snowLands = distributeGroup(groupByName(data.snowLands));

	while (data.baseLands.length || data.snowLands.length || data.others.length)
	{

		for (let i = 0; i < 4; i++)
		{
			let card = data.others.pop();
			arr[i].push(card);

			if (!card || (idx++ % 4) === 0)
			{
				arr[i].push(data.baseLands.pop() ?? data.snowLands.pop())
				arr[i].push(data.baseLands.pop() ?? data.snowLands.pop())
			}
		}

	}

	return arr.flat().filter(c => c)
}

export default distributeCards
