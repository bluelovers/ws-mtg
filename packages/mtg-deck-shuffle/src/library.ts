import { DeckLibrary, ICardOfLibrary } from 'mtg-decklist-to-library';
import { IOptionsDeckLibraryWithShuffle } from './types';
import { getRandomFromOptions } from './util/rand';
import { splitThenMerge } from './method/splitThenMerge';
import { Decklist } from 'mtg-decklist-parser2';
import { distributeCards } from './method/distributeCards';

export class DeckLibraryWithShuffle<T = {}> extends DeckLibrary<ICardOfLibrary<T>>
{

	public _options?: IOptionsDeckLibraryWithShuffle<ICardOfLibrary<T>>;

	constructor(deck: Decklist, options?: IOptionsDeckLibraryWithShuffle<ICardOfLibrary<T>>)
	{
		super(deck, options);
	}

	options(options?: IOptionsDeckLibraryWithShuffle): IOptionsDeckLibraryWithShuffle
	{
		return {
			...this._options,
			...options,
		}
	}

	override shuffleStarting()
	{
		let options = this.options();

		this.cards = [
			distributeCards,
			splitThenMerge,
		].reduce((cards, fn) => fn(cards, options), this.cards);

		this._shuffleStarting = true;
	}

	override shuffle(isStarting?: boolean)
	{
		let random = getRandomFromOptions(this.options());

		if (isStarting)
		{

		}

		random.arrayShuffle(this.cards, true);
	}

}
