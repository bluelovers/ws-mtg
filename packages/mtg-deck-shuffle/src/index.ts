export { IOptionsDeckLibraryWithShuffle } from './types';

import { DeckLibraryWithShuffle } from './library';

export { distributeCards } from './method/distributeCards';
export { splitThenMerge } from './method/splitThenMerge';
export { ensureLands } from './method/ensureLands';

export { DeckLibraryWithShuffle }

export default DeckLibraryWithShuffle
