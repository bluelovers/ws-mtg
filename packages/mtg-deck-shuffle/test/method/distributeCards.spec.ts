import { readFile } from 'fs-extra';
import { Decklist } from 'mtg-decklist-parser2';
import { DeckLibraryWithShuffle } from '../../src/library';
import { ICardWithoutAmount } from 'mtg-decklist-parser2';
import { join } from 'path';
import { __root } from '../__root';
import { groupByName } from '../../src/util/groupByName';
import { distributeGroup } from '../../src/util/distributeGroup';
import distributeCards from '../../src/method/distributeCards';
import { sumGroupCards } from '../../src/util/sumGroupCards';
import { filterLands } from '../../src/util/filterLands';

describe(`distributeCards`, () =>
{

	let cards: ICardWithoutAmount[];

	beforeEach(async () =>
	{
		cards = await readFile(join(__root, 'test', 'deck/whiteline.txt'))
			.then(rawString => new Decklist(rawString))
			.then(parsed =>
			{
				return new DeckLibraryWithShuffle(parsed).cards
			})
	});

	test(`groupByName`, () =>
	{

		let actual = groupByName(cards);

		expect(actual).toMatchSnapshot();

		expect(sumGroupCards(actual)).toStrictEqual(cards.length);

		let actual2 = distributeGroup(actual);

		expect(actual2).toMatchSnapshot();
		expect(actual2).toHaveProperty('length', cards.length);

	});

	test(`distributeCards`, () =>
	{

		let actual = distributeCards(cards);

		expect(actual).toMatchSnapshot();
		expect(actual).toHaveProperty('length', cards.length);

	});

	test(`distribute:baseLands`, () =>
	{
		let data = filterLands(cards);

		let actual = distributeGroup(groupByName(data.baseLands));

		expect(actual).toMatchSnapshot();

	});

	test(`distribute:snowLands`, () =>
	{
		let data = filterLands(cards);

		let actual = distributeGroup(groupByName(data.snowLands));

		expect(actual).toMatchSnapshot();

	});

	test(`distribute:others`, () =>
	{
		let data = filterLands(cards);

		let actual = distributeGroup(groupByName(data.others));

		expect(actual).toMatchSnapshot();

	});

	test(`distribute:check`, () =>
	{
		let data = [
			{
				"name": "Forest",
			},
			{
				"name": "Plains",
			},
			{
				"name": "Forest",
			},
			{
				"name": "Plains",
			},
			{
				"name": "Forest",
			},
			{
				"name": "Plains",
			},
		];

		let actual = distributeGroup(groupByName(data));

		expect(actual).toMatchSnapshot();
		expect(actual).toStrictEqual(data);

	});

});
