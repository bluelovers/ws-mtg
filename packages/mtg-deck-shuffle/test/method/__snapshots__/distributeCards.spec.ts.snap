// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distributeCards distributeCards 1`] = `
Array [
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Revitalize",
    "set": "KHM",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Pillardrop Rescuer",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 126,
    "mtgoID": undefined,
    "name": "Devouring Tendrils",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 132,
    "mtgoID": undefined,
    "name": "Fortifying Draught",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 215,
    "mtgoID": undefined,
    "name": "You Meet in a Tavern",
    "set": "AFR",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 20,
    "mtgoID": undefined,
    "name": "Icingdeath, Frost Tyrant",
    "set": "AFR",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 22,
    "mtgoID": undefined,
    "name": "Kor Celebrant",
    "set": "ZNR",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 5,
    "mtgoID": undefined,
    "name": "Celestial Unicorn",
    "set": "AFR",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 200,
    "mtgoID": undefined,
    "name": "Prosperous Innkeeper",
    "set": "AFR",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 236,
    "mtgoID": undefined,
    "name": "Trelasarra, Moon Dancer",
    "set": "AFR",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Revitalize",
    "set": "KHM",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Pillardrop Rescuer",
    "set": "STX",
  },
  Object {
    "collectors": 24,
    "mtgoID": undefined,
    "name": "Minimus Containment",
    "set": "AFR",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 258,
    "mtgoID": undefined,
    "name": "Branchloft Pathway",
    "set": "ZNR",
  },
  Object {
    "collectors": 215,
    "mtgoID": undefined,
    "name": "You Meet in a Tavern",
    "set": "AFR",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Valkyrie Harbinger",
    "set": "KHM",
  },
  Object {
    "collectors": 22,
    "mtgoID": undefined,
    "name": "Kor Celebrant",
    "set": "ZNR",
  },
  Object {
    "collectors": 5,
    "mtgoID": undefined,
    "name": "Celestial Unicorn",
    "set": "AFR",
  },
  Object {
    "collectors": 200,
    "mtgoID": undefined,
    "name": "Prosperous Innkeeper",
    "set": "AFR",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Revitalize",
    "set": "KHM",
  },
  Object {
    "collectors": 126,
    "mtgoID": undefined,
    "name": "Devouring Tendrils",
    "set": "STX",
  },
  Object {
    "collectors": 24,
    "mtgoID": undefined,
    "name": "Minimus Containment",
    "set": "AFR",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 220,
    "mtgoID": undefined,
    "name": "Drizzt Do'Urden",
    "set": "AFR",
  },
  Object {
    "collectors": 194,
    "mtgoID": undefined,
    "name": "Lurking Roper",
    "set": "AFR",
  },
  Object {
    "collectors": 6,
    "mtgoID": undefined,
    "name": "Cleric Class",
    "set": "AFR",
  },
  Object {
    "collectors": 22,
    "mtgoID": undefined,
    "name": "Kor Celebrant",
    "set": "ZNR",
  },
  Object {
    "collectors": 34,
    "mtgoID": undefined,
    "name": "Story Seeker",
    "set": "KHM",
  },
  Object {
    "collectors": 24,
    "mtgoID": undefined,
    "name": "Righteous Valkyrie",
    "set": "KHM",
  },
  Object {
    "collectors": 119,
    "mtgoID": undefined,
    "name": "Accomplished Alchemist",
    "set": "STX",
  },
  Object {
    "collectors": 126,
    "mtgoID": undefined,
    "name": "Devouring Tendrils",
    "set": "STX",
  },
  Object {
    "collectors": 132,
    "mtgoID": undefined,
    "name": "Fortifying Draught",
    "set": "STX",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 186,
    "mtgoID": undefined,
    "name": "Inscription of Abundance",
    "set": "ZNR",
  },
  Object {
    "collectors": 194,
    "mtgoID": undefined,
    "name": "Lurking Roper",
    "set": "AFR",
  },
  Object {
    "collectors": 6,
    "mtgoID": undefined,
    "name": "Cleric Class",
    "set": "AFR",
  },
  Object {
    "collectors": 5,
    "mtgoID": undefined,
    "name": "Celestial Unicorn",
    "set": "AFR",
  },
  Object {
    "collectors": 34,
    "mtgoID": undefined,
    "name": "Story Seeker",
    "set": "KHM",
  },
  Object {
    "collectors": 236,
    "mtgoID": undefined,
    "name": "Trelasarra, Moon Dancer",
    "set": "AFR",
  },
]
`;

exports[`distributeCards groupByName 1`] = `
Object {
  "Accomplished Alchemist": Array [
    Object {
      "collectors": 119,
      "mtgoID": undefined,
      "name": "Accomplished Alchemist",
      "set": "STX",
    },
  ],
  "Arctic Treeline": Array [
    Object {
      "collectors": 249,
      "mtgoID": undefined,
      "name": "Arctic Treeline",
      "set": "KHM",
    },
    Object {
      "collectors": 249,
      "mtgoID": undefined,
      "name": "Arctic Treeline",
      "set": "KHM",
    },
    Object {
      "collectors": 249,
      "mtgoID": undefined,
      "name": "Arctic Treeline",
      "set": "KHM",
    },
    Object {
      "collectors": 249,
      "mtgoID": undefined,
      "name": "Arctic Treeline",
      "set": "KHM",
    },
  ],
  "Branchloft Pathway": Array [
    Object {
      "collectors": 258,
      "mtgoID": undefined,
      "name": "Branchloft Pathway",
      "set": "ZNR",
    },
  ],
  "Celestial Unicorn": Array [
    Object {
      "collectors": 5,
      "mtgoID": undefined,
      "name": "Celestial Unicorn",
      "set": "AFR",
    },
    Object {
      "collectors": 5,
      "mtgoID": undefined,
      "name": "Celestial Unicorn",
      "set": "AFR",
    },
    Object {
      "collectors": 5,
      "mtgoID": undefined,
      "name": "Celestial Unicorn",
      "set": "AFR",
    },
  ],
  "Cleric Class": Array [
    Object {
      "collectors": 6,
      "mtgoID": undefined,
      "name": "Cleric Class",
      "set": "AFR",
    },
    Object {
      "collectors": 6,
      "mtgoID": undefined,
      "name": "Cleric Class",
      "set": "AFR",
    },
  ],
  "Devouring Tendrils": Array [
    Object {
      "collectors": 126,
      "mtgoID": undefined,
      "name": "Devouring Tendrils",
      "set": "STX",
    },
    Object {
      "collectors": 126,
      "mtgoID": undefined,
      "name": "Devouring Tendrils",
      "set": "STX",
    },
    Object {
      "collectors": 126,
      "mtgoID": undefined,
      "name": "Devouring Tendrils",
      "set": "STX",
    },
  ],
  "Drizzt Do'Urden": Array [
    Object {
      "collectors": 220,
      "mtgoID": undefined,
      "name": "Drizzt Do'Urden",
      "set": "AFR",
    },
  ],
  "Forest": Array [
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Forest",
      "set": "STX",
    },
  ],
  "Fortifying Draught": Array [
    Object {
      "collectors": 132,
      "mtgoID": undefined,
      "name": "Fortifying Draught",
      "set": "STX",
    },
    Object {
      "collectors": 132,
      "mtgoID": undefined,
      "name": "Fortifying Draught",
      "set": "STX",
    },
  ],
  "Icingdeath, Frost Tyrant": Array [
    Object {
      "collectors": 20,
      "mtgoID": undefined,
      "name": "Icingdeath, Frost Tyrant",
      "set": "AFR",
    },
  ],
  "Inscription of Abundance": Array [
    Object {
      "collectors": 186,
      "mtgoID": undefined,
      "name": "Inscription of Abundance",
      "set": "ZNR",
    },
  ],
  "Kor Celebrant": Array [
    Object {
      "collectors": 22,
      "mtgoID": undefined,
      "name": "Kor Celebrant",
      "set": "ZNR",
    },
    Object {
      "collectors": 22,
      "mtgoID": undefined,
      "name": "Kor Celebrant",
      "set": "ZNR",
    },
    Object {
      "collectors": 22,
      "mtgoID": undefined,
      "name": "Kor Celebrant",
      "set": "ZNR",
    },
  ],
  "Lurking Roper": Array [
    Object {
      "collectors": 194,
      "mtgoID": undefined,
      "name": "Lurking Roper",
      "set": "AFR",
    },
    Object {
      "collectors": 194,
      "mtgoID": undefined,
      "name": "Lurking Roper",
      "set": "AFR",
    },
  ],
  "Minimus Containment": Array [
    Object {
      "collectors": 24,
      "mtgoID": undefined,
      "name": "Minimus Containment",
      "set": "AFR",
    },
    Object {
      "collectors": 24,
      "mtgoID": undefined,
      "name": "Minimus Containment",
      "set": "AFR",
    },
  ],
  "Pillardrop Rescuer": Array [
    Object {
      "collectors": 23,
      "mtgoID": undefined,
      "name": "Pillardrop Rescuer",
      "set": "STX",
    },
    Object {
      "collectors": 23,
      "mtgoID": undefined,
      "name": "Pillardrop Rescuer",
      "set": "STX",
    },
  ],
  "Plains": Array [
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
    Object {
      "collectors": 367,
      "mtgoID": undefined,
      "name": "Plains",
      "set": "STX",
    },
  ],
  "Prosperous Innkeeper": Array [
    Object {
      "collectors": 200,
      "mtgoID": undefined,
      "name": "Prosperous Innkeeper",
      "set": "AFR",
    },
    Object {
      "collectors": 200,
      "mtgoID": undefined,
      "name": "Prosperous Innkeeper",
      "set": "AFR",
    },
  ],
  "Revitalize": Array [
    Object {
      "collectors": 23,
      "mtgoID": undefined,
      "name": "Revitalize",
      "set": "KHM",
    },
    Object {
      "collectors": 23,
      "mtgoID": undefined,
      "name": "Revitalize",
      "set": "KHM",
    },
    Object {
      "collectors": 23,
      "mtgoID": undefined,
      "name": "Revitalize",
      "set": "KHM",
    },
  ],
  "Righteous Valkyrie": Array [
    Object {
      "collectors": 24,
      "mtgoID": undefined,
      "name": "Righteous Valkyrie",
      "set": "KHM",
    },
  ],
  "Story Seeker": Array [
    Object {
      "collectors": 34,
      "mtgoID": undefined,
      "name": "Story Seeker",
      "set": "KHM",
    },
    Object {
      "collectors": 34,
      "mtgoID": undefined,
      "name": "Story Seeker",
      "set": "KHM",
    },
  ],
  "Trelasarra, Moon Dancer": Array [
    Object {
      "collectors": 236,
      "mtgoID": undefined,
      "name": "Trelasarra, Moon Dancer",
      "set": "AFR",
    },
    Object {
      "collectors": 236,
      "mtgoID": undefined,
      "name": "Trelasarra, Moon Dancer",
      "set": "AFR",
    },
  ],
  "Valkyrie Harbinger": Array [
    Object {
      "collectors": 374,
      "mtgoID": undefined,
      "name": "Valkyrie Harbinger",
      "set": "KHM",
    },
  ],
  "You Meet in a Tavern": Array [
    Object {
      "collectors": 215,
      "mtgoID": undefined,
      "name": "You Meet in a Tavern",
      "set": "AFR",
    },
    Object {
      "collectors": 215,
      "mtgoID": undefined,
      "name": "You Meet in a Tavern",
      "set": "AFR",
    },
  ],
}
`;

exports[`distributeCards groupByName 2`] = `
Array [
  Object {
    "collectors": 236,
    "mtgoID": undefined,
    "name": "Trelasarra, Moon Dancer",
    "set": "AFR",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Valkyrie Harbinger",
    "set": "KHM",
  },
  Object {
    "collectors": 186,
    "mtgoID": undefined,
    "name": "Inscription of Abundance",
    "set": "ZNR",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Pillardrop Rescuer",
    "set": "STX",
  },
  Object {
    "collectors": 200,
    "mtgoID": undefined,
    "name": "Prosperous Innkeeper",
    "set": "AFR",
  },
  Object {
    "collectors": 22,
    "mtgoID": undefined,
    "name": "Kor Celebrant",
    "set": "ZNR",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 126,
    "mtgoID": undefined,
    "name": "Devouring Tendrils",
    "set": "STX",
  },
  Object {
    "collectors": 5,
    "mtgoID": undefined,
    "name": "Celestial Unicorn",
    "set": "AFR",
  },
  Object {
    "collectors": 126,
    "mtgoID": undefined,
    "name": "Devouring Tendrils",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 24,
    "mtgoID": undefined,
    "name": "Righteous Valkyrie",
    "set": "KHM",
  },
  Object {
    "collectors": 5,
    "mtgoID": undefined,
    "name": "Celestial Unicorn",
    "set": "AFR",
  },
  Object {
    "collectors": 20,
    "mtgoID": undefined,
    "name": "Icingdeath, Frost Tyrant",
    "set": "AFR",
  },
  Object {
    "collectors": 220,
    "mtgoID": undefined,
    "name": "Drizzt Do'Urden",
    "set": "AFR",
  },
  Object {
    "collectors": 132,
    "mtgoID": undefined,
    "name": "Fortifying Draught",
    "set": "STX",
  },
  Object {
    "collectors": 119,
    "mtgoID": undefined,
    "name": "Accomplished Alchemist",
    "set": "STX",
  },
  Object {
    "collectors": 34,
    "mtgoID": undefined,
    "name": "Story Seeker",
    "set": "KHM",
  },
  Object {
    "collectors": 6,
    "mtgoID": undefined,
    "name": "Cleric Class",
    "set": "AFR",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Pillardrop Rescuer",
    "set": "STX",
  },
  Object {
    "collectors": 22,
    "mtgoID": undefined,
    "name": "Kor Celebrant",
    "set": "ZNR",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Revitalize",
    "set": "KHM",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 200,
    "mtgoID": undefined,
    "name": "Prosperous Innkeeper",
    "set": "AFR",
  },
  Object {
    "collectors": 22,
    "mtgoID": undefined,
    "name": "Kor Celebrant",
    "set": "ZNR",
  },
  Object {
    "collectors": 194,
    "mtgoID": undefined,
    "name": "Lurking Roper",
    "set": "AFR",
  },
  Object {
    "collectors": 258,
    "mtgoID": undefined,
    "name": "Branchloft Pathway",
    "set": "ZNR",
  },
  Object {
    "collectors": 24,
    "mtgoID": undefined,
    "name": "Minimus Containment",
    "set": "AFR",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Revitalize",
    "set": "KHM",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 194,
    "mtgoID": undefined,
    "name": "Lurking Roper",
    "set": "AFR",
  },
  Object {
    "collectors": 132,
    "mtgoID": undefined,
    "name": "Fortifying Draught",
    "set": "STX",
  },
  Object {
    "collectors": 23,
    "mtgoID": undefined,
    "name": "Revitalize",
    "set": "KHM",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 34,
    "mtgoID": undefined,
    "name": "Story Seeker",
    "set": "KHM",
  },
  Object {
    "collectors": 6,
    "mtgoID": undefined,
    "name": "Cleric Class",
    "set": "AFR",
  },
  Object {
    "collectors": 215,
    "mtgoID": undefined,
    "name": "You Meet in a Tavern",
    "set": "AFR",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 126,
    "mtgoID": undefined,
    "name": "Devouring Tendrils",
    "set": "STX",
  },
  Object {
    "collectors": 236,
    "mtgoID": undefined,
    "name": "Trelasarra, Moon Dancer",
    "set": "AFR",
  },
  Object {
    "collectors": 5,
    "mtgoID": undefined,
    "name": "Celestial Unicorn",
    "set": "AFR",
  },
  Object {
    "collectors": 215,
    "mtgoID": undefined,
    "name": "You Meet in a Tavern",
    "set": "AFR",
  },
  Object {
    "collectors": 24,
    "mtgoID": undefined,
    "name": "Minimus Containment",
    "set": "AFR",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
  Object {
    "collectors": 249,
    "mtgoID": undefined,
    "name": "Arctic Treeline",
    "set": "KHM",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 374,
    "mtgoID": undefined,
    "name": "Forest",
    "set": "STX",
  },
  Object {
    "collectors": 367,
    "mtgoID": undefined,
    "name": "Plains",
    "set": "STX",
  },
]
`;
